pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,283 @@
+		AgentHostname:  a.agentHostname,
+		AgentVersion:   info.Version,
+		ClientComputed: true,
+	}
+}
+
+type bucket struct {
+	// first is the first payload matching the bucket. If a second payload matches the bucket
+	// this field will be empty
+	first pb.ClientStatsPayload
+	// ts is the unix timestamp attached to the payload
+	ts uint64
+	// n counts the number of payloads matching the bucket
+	n int
+	// agg contains the aggregated Hits/Errors/Duration counts
+	agg map[payloadAggregationKey]map[bucketAggregationKey]*aggregatedCounts
+}
+
+func (b *bucket) add(p pb.ClientStatsPayload) []pb.ClientStatsPayload {
+	b.n++
+	if b.n == 1 {
+		b.first = p
+		return nil
+	}
+	// if it's the second payload we flush the first payload with counts trimmed
+	if b.n == 2 {
+		first := b.first
+		b.first = pb.ClientStatsPayload{}
+		b.agg = make(map[payloadAggregationKey]map[bucketAggregationKey]*aggregatedCounts, 2)
+		b.aggregateCounts(first)
+		b.aggregateCounts(p)
+		first.AgentAggregation = distributionsPayload
+		p.AgentAggregation = distributionsPayload
+		return []pb.ClientStatsPayload{trimHitsErrors(first), trimHitsErrors(p)}
+	}
+	b.aggregateCounts(p)
+	p.AgentAggregation = distributionsPayload
+	return []pb.ClientStatsPayload{trimHitsErrors(p)}
+}
+
+func (b *bucket) aggregateCounts(p pb.ClientStatsPayload) {
+	payloadAggKey := newPayloadAggregationKey(p.Env, p.Hostname, p.Version)
+	payloadAgg, ok := b.agg[payloadAggKey]
+	if !ok {
+		var size int
+		for _, s := range p.Stats {
+			size += len(s.Stats)
+		}
+		payloadAgg = make(map[bucketAggregationKey]*aggregatedCounts, size)
+		b.agg[payloadAggKey] = payloadAgg
+	}
+	for _, s := range p.Stats {
+		for _, sb := range s.Stats {
+			aggKey := newBucketAggregationKey(sb)
+			agg, ok := payloadAgg[aggKey]
+			if !ok {
+				agg = &aggregatedCounts{}
+				payloadAgg[aggKey] = agg
+			}
+			agg.hits += sb.Hits
+			agg.errors += sb.Errors
+			agg.duration += sb.Duration
+		}
+	}
+}
+
+func (b *bucket) flush() []pb.ClientStatsPayload {
+	if b.n == 1 {
+		return []pb.ClientStatsPayload{b.first}
+	}
+	return b.aggregationToPayloads()
+}
+
+func (b *bucket) aggregationToPayloads() []pb.ClientStatsPayload {
+	res := make([]pb.ClientStatsPayload, 0, len(b.agg))
+	for payloadKey, aggrCounts := range b.agg {
+		stats := make([]pb.ClientGroupedStats, 0, len(aggrCounts))
+		for aggrKey, counts := range aggrCounts {
+			stats = append(stats, pb.ClientGroupedStats{
+				Service:        aggrKey.service,
+				Name:           aggrKey.name,
+				Resource:       aggrKey.resource,
+				HTTPStatusCode: aggrKey.statusCode,
+				Type:           aggrKey.typ,
+				Hits:           counts.hits,
+				Errors:         counts.errors,
+				Duration:       counts.duration,
+			})
+		}
+		clientBuckets := []pb.ClientStatsBucket{
+			{
+				Start:    b.ts * 1e9, // to nanosecond
+				Duration: bucketDuration,
+				Stats:    stats,
+			}}
+		res = append(res, pb.ClientStatsPayload{
+			Hostname:         payloadKey.hostname,
```

Comment:
Does it have an impact that we are missing Lang, TracerVersion, RuntimeID, Sequence?
Are these fields used anywhere?

Reply:
No these fields are not currently used or aggregated on. 
They lose their meaning on the payload aggregating counts as multiple tracers/versions/langs/runtimeID/sequence could be involved